### PR TITLE
Feature json support

### PR DESCRIPTION
**Pull request by Stephanevg**
## 1. General summary of the pull request

Hi,

This PR is to add support to Export the PesterObject to JSON format. I discussed it in this issue here -> https://github.com/pester/Pester/issues/1322

This would be first draft (I guess) as I still need to write the tests for this. (But I'll do that, once we agree on the Implementation).
Please let me know if everything is ok with this PR, or if I should adapt it. (Stuff like location etc..).

# Testing

I used the code below to test it:

```Powershell

import-module ./Pester/Pester.psd1 -force
$p = invoke-pester ./Pester/Pester.Tests.ps1  -OutputFormat JSON -OutputFile ./plop.json -PassThru
$p
(gc './plop.json' -Raw) | ConvertFrom-Json


```

These are the results I get (The objects are identical, except for the order of the hash tables, but that is something we cannot workaround as Pester is backwards compatible with PS 2.0 and `[ordered]` hash tables are only available since PS3.

Capture d’écran 2019-07-26 à 01 44 21

### CODE REVIEW

File: Functions/Get-ShouldOperator.ps1 [hunk 1]
```diff
@@ -65,12 +65,14 @@ function Get-ShouldOperator {
 
     END {
         if ($Name) {
+
```
Comment (nohwnd) [inline Functions/Get-ShouldOperator.ps1]: This file contains just unnecesary whitespace changes, please revert.

File: Functions/TestResults.ps1 [hunk 2]
```diff
@@ -33,12 +33,21 @@ function Export-PesterResults {
         [string] $Format
     )
 
-    switch ($Format) {
-        'NUnitXml' {
-            Export-NUnitReport -PesterState $PesterState -Path $Path
+
+    switch ($Format)
```
Comment (nohwnd) [inline Functions/TestResults.ps1]: Please use Stroustrup formatting as the rest of the code to avoid unnecessary changes.

File: Functions/Get-ShouldOperator.ps1 (hunk 1 above)
Comment (Stephanevg) [reply Functions/Get-ShouldOperator.ps1]: This has been removed in last commit

File: Functions/TestResults.ps1 (hunk 2 above)
Comment (Stephanevg) [reply Functions/TestResults.ps1]: Ok, should also be fixed. 
image

**Comment by Stephanevg** [comment]
Tests will come later.

**Comment by Stephanevg** [comment]
So, I have created the needed Pester Tests for this feature.
The JSON output file looks like this one:

```json
{
  "TagFilter": null,
  "ExcludeTagFilter": null,
  "TestNameFilter": null,
  "ScriptBlockFilter": null,
  "TotalCount": 2,
  "PassedCount": 1,
  "FailedCount": 1,
  "SkippedCount": 0,
  "PendingCount": 0,
  "InconclusiveCount": 0,
  "Time": {
    "Ticks": 0,
    "Days": 0,
    "Hours": 0,
    "Milliseconds": 0,
    "Minutes": 0,
    "Seconds": 0,
    "TotalDays": 0.0,
    "TotalHours": 0.0,
    "TotalMilliseconds": 0.0,
    "TotalMinutes": 0.0,
    "TotalSeconds": 0.0
  },
  "TestResult": [
    {
      "Describe": "Fake Describe Block",
      "StackTrace": "",
      "Parameters": null,
      "Time": "00:00:01",
      "Passed": true,
      "Show": 1023,
      "Context": "",
      "ParameterizedSuiteName": "",
      "Result": "Passed",
      "FailureMessage": "",
      "Name": "Successful testcase",
      "ErrorRecord": null
    },
    {
      "Describe": "Fake Describe Block",
      "StackTrace": "",
      "Parameters": null,
      "Time": "00:00:02",
      "Passed": false,
      "Show": 1023,
      "Context": "",
      "ParameterizedSuiteName": "",
      "Result": "Failed",
      "FailureMessage": "",
      "Name": "Failed testcase",
      "ErrorRecord": null
    }
  ]
}

```

Tests on my side (On a Mac) show all green:

Capture d’écran 2019-08-02 à 20 43 53

**Comment by Stephanevg** [comment]
Ach, ok looking at the test results, I see that the tests do pass on travis (meaning MacOS and Linux - thus Ps5.1 or above) but not on Windows. I specifically think the tests fail on ps2.0 boxes
```powershell
Export to JSON
    [-] Should not throw when JSON export is select 14ms
      Expected no exception to be thrown, but an exception "The term 'ConvertTo-Json' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again." was thrown from C:\agent\_work\3\s\Functions\TestResults.ps1:46 char:106
          +             $PesterState | & $script:SafeCommands['Select-Object'] -Property $properties | ConvertTo-Json <<<<  | Out-File -Path $Path -Encoding utf8.
      349:             { Export-PesterResults -pesterState $testResults -Path $testFile -Format JSON } | Should not throw
      at line: 349 in C:\agent\_work\3\s\Functions\TestResults.Tests.ps1
    [-] Should create a .json file 6ms
      CommandNotFoundException: The term 'ConvertTo-Json' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
      at line: 46 in C:\agent\_work\3\s\Functions\TestResults.ps1

    .JSON File should contain correct data
      [-] Error occurred in Context block 0ms
        CommandNotFoundException: The term 'ConvertTo-Json' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
        at line: 46 in C:\agent\_work\3\s\Functions\TestResults.ps1
```

I actually forgot, but Convertto-Json is available on PS version 3 and upwards :(
I think at PsConfEU @nohwnd talked about removing support from ps2 in v5 of Pester. If so, than this feature might be best suited for PS5.

**Comment by nohwnd** [comment]
If you are still interested in adding this, please fix the merge conflicts and then we can talk. But is this really necessary when you have the result object and ConvertTo-Json cmdlet?

**Comment by Stephanevg** [comment]
> If you are still interested in adding this, please fix the merge conflicts and then we can talk. But is this really necessary when you have the result object and ConvertTo-Json cmdlet?

wow, this one goes a while back. I actually think I am. 
My whole point of the PR was that it would be nice to have a 'standard' and 'official' way to dehydrate the pester object. 
More as a convenvience for the engineer that has to use pester.

Ultimatley the format doesn't matter to be honnest (I have actually built a wrapper at work that exports to CLIXML, as I had some issues with JSON). This function could easily be updated for that format - if desired.

Should i push against the same branch again or change it to another one?

**Comment by nohwnd** [comment]
I'd prefer clixml in that case. 

Branching is up to you, either use this branch, or start a new PR. If you will use clixml there probably won't be any common code with this PR so it might be easier to abandon it.
